pubsub/producer.go

```diff
@@ -102,150 +87,88 @@ func NewProducer[Request any, Response any](client redis.UniversalClient, stream
-			continue
-		}
-		var req Request
-		if err := json.Unmarshal([]byte(data), &req); err != nil {
-			log.Error("redis producer reproduce: message not a request", "id", msg.ID, "err", err, "value", msg.Values[messageKey])
-			continue
-		}
-		if _, err := p.client.XAck(ctx, p.redisStream, p.redisGroup, msg.ID).Result(); err != nil {
-			log.Error("redis producer reproduce: could not ACK", "id", msg.ID, "err", err)
-			continue
-		}
-		// Only re-insert messages that were removed the the pending list first.
-		if _, err := p.reproduce(ctx, req, msg.ID); err != nil {
-			log.Error("redis producer reproduce: error", "err", err)
-		}
-	}
-	return nil
-}
-
-func setMinIdInt(min *[2]uint64, id string) error {
-	idParts := strings.Split(id, "-")
+func setMaxMsgIdInt(maxMsgIdInt *[2]uint64, msgId string) error {
+	idParts := strings.Split(msgId, "-")
 	if len(idParts) != 2 {
-		return fmt.Errorf("invalid i.d: %v", id)
+		return fmt.Errorf("invalid i.d: %v", msgId)
 	}
 	idTimeStamp, err := strconv.ParseUint(idParts[0], 10, 64)
 	if err != nil {
-		return fmt.Errorf("invalid i.d: %v err: %w", id, err)
+		return fmt.Errorf("invalid i.d: %v err: %w", msgId, err)
 	}
-	if idTimeStamp > min[0] {
+	if idTimeStamp < maxMsgIdInt[0] {
 		return nil
 	}
 	idSerial, err := strconv.ParseUint(idParts[1], 10, 64)
 	if err != nil {
-		return fmt.Errorf("invalid i.d serial: %v err: %w", id, err)
+		return fmt.Errorf("invalid i.d serial: %v err: %w", msgId, err)
 	}
-	if idTimeStamp < min[0] {
-		min[0] = idTimeStamp
-		min[1] = idSerial
+	if idTimeStamp > maxMsgIdInt[0] {
+		maxMsgIdInt[0] = idTimeStamp
+		maxMsgIdInt[1] = idSerial
 		return nil
 	}
-	// idTimeStamp == min[0]
-	if idSerial < min[1] {
-		min[1] = idSerial
+	// idTimeStamp == maxMsgIdInt[0]
+	if idSerial > maxMsgIdInt[1] {
+		maxMsgIdInt[1] = idSerial
 	}
 	return nil
 }
 
 // checkResponses checks iteratively whether response for the promise is ready.
 func (p *Producer[Request, Response]) checkResponses(ctx context.Context) time.Duration {
-	minIdInt := [2]uint64{math.MaxUint64, math.MaxUint64}
+	maxMsgIdInt := [2]uint64{0, 0}
 	p.promisesLock.Lock()
 	defer p.promisesLock.Unlock()
 	responded := 0
 	errored := 0
-	for id, promise := range p.promises {
+	for id, msgIDAndPromise := range p.promises {
 		if ctx.Err() != nil {
 			return 0
 		}
-		res, err := p.client.Get(ctx, id).Result()
+		msgKey := MessageKeyFor(p.redisStream, id)
+		res, err := p.client.Get(ctx, msgKey).Result()
 		if err != nil {
-			errSetId := setMinIdInt(&minIdInt, id)
-			if errSetId != nil {
-				log.Error("error setting minId", "err", err)
-				return p.cfg.CheckResultInterval
-			}
 			if !errors.Is(err, redis.Nil) {
 				log.Error("Error reading value in redis", "key", id, "error", err)
 			}
 			continue
 		}
+		// We keep track of a maxMsgId of a successfully solved request, because messages
+		// with id lower than this are either ack-ed or in PEL, so its safe to call XTRIMMINID on maxMsgId
```

Comment:
* maybe we could get a minimal redis id (min_id) from PEL (eg. using XPENDING) and then XTRIM (min_id - 1) ? or maybe we could figure out min_id base on data that we already have?

* alternatively, maybe we should just remove msgs for solved requests with XDEL? looks like it could also be efficient, as `"normally what happens when you delete an entry from a stream is that the entry is not really evicted, it just gets marked as deleted. Eventually if all the entries in a macro-node are marked as deleted, the whole node is destroyed and the memory reclaimed."` (https://redis.io/docs/latest/commands/xdel/)

Reply:
I thought it would be OK to trim messages that are in PEL because I was under the impression that these messages would persist in the PEL, they do persist, but very frustratingly only their IDs presist :( kinda like ghost items. I just found out through this [thread](https://github.com/redis/redis/issues/6199#issuecomment-516344848). 

Unfortunately, there is really no meaningful way to take advantage of such a design and there is also an ongoing popular feature request for [XTRIM ](https://github.com/redis/redis/issues/5774) to only delete messages that are acked. 

The solution you pointed out about using xtrimminid with the minima of PEL was also pointed out by @tsahee and seems like it should work out- along with a way to handle old messages that might prevent xtriming from happening altogether. Consumers calling XDEL after acking a message might work too, but this has data fragmentation issue as messages arent really evicted from the streams- pointed out [here](https://redis.io/docs/latest/commands/xdel/), 

> Redis streams are represented in a way that makes them memory efficient: a radix tree is used in order to index macro-nodes that pack linearly tens of stream entries. Normally what happens when you delete an entry from a stream is that the entry is not really evicted, it just gets marked as deleted.
...
Eventually if all the entries in a macro-node are marked as deleted, the whole node is destroyed and the memory reclaimed. This means that if you delete a large amount of entries from a stream, for instance more than 50% of the entries appended to the stream, the memory usage per entry may increment, since what happens is that the stream will become fragmented. However the stream performance will remain the same.

**I think a good middle ground is for every producer to - [trim using xtrimminid on PEL's least id] + [look at PEL's least id and find out when it was first added to the stream (first part of its message id) and if that is longer than a certain duration from now, producer will ack it and subsequently will also error the promise having that message if it originated from this producer]. let me know what you think about this @tsahee, @magicxyyz** 
 
Thanks for catching this, the tests were unable to detect this as PEL is populated in the beginning and that would then be xautoclaimed by other consumers before xtrim was called

Reply:
I'll take a deeper look into that later, just a quick comment here to the use of XDEL: 

I think that in our case XDEL won't create persisting fragmentation only temporary one that shouldn't use much more memory then if we trim messages with ID lower then minimal ID from PEL. We are processing the messages in IDs order, so if we can XTRIM messages up to some ID then it means that we already finished processing them => we can as well XDEL them after finishing processing them and they will eventually be removed with whole macro-nodes. That should be the same effect as if you could use ["Nearly exact trimming" ](https://redis.io/docs/latest/commands/xtrim/)with MINID.